### PR TITLE
175 factory add query displaying data for given user

### PR DESCRIPTION
**Pull request by gangov**

### CODE REVIEW

File: contracts/factory/src/contract.rs [hunk 1]
```diff
@@ -168,6 +170,45 @@ impl FactoryTrait for Factory {
...
+                &Symbol::new(&env, "query_pool_info_for_factory"),
+                Vec::new(&env),
+            );
+            // todo here we check if the lp balance is over 0
+            if response.pool_response.asset_lp_share.amount > 0 {
+                lp_share_info.push_back(LpShareInfo {
+                    owner: caller.clone(),
+                    amount: response.pool_response.asset_lp_share.amount,
+                })
+            }
```
Comment (ueco-jb) [inline contracts/factory/src/contract.rs:191]: You iterate over all LP addresses, you query each pool for that pool's data (not the user's!) and then you check if each pool has `lp_share` amount bigger then 0... And you assign that general amount to a user. This is invalid.

File: contracts/factory/src/storage.rs [hunk 2]
```diff
@@ -30,6 +30,27 @@ pub struct Config {
     pub multihop_address: Address,
 }
 
+#[contracttype]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LpAndStakeInfo {
```
Comment (ueco-jb) [inline contracts/factory/src/storage.rs:35]: How about:
```suggestion
pub struct UserPortfolio {
```
the rest can stay as it is.

**Review by ueco-jb — COMMENTED** [review]
Good progress.

File: contracts/pool_stable/src/contract.rs [hunk 3]
```diff
@@ -500,7 +500,7 @@ impl StableLiquidityPoolTrait for StableLiquidityPool {
                 amount: utils::get_total_shares(&env),
             },
         };
-        let total_fee_bps = config.max_allowed_spread_bps;
+        let total_fee_bps = config.total_fee_bps;
```
Comment (ueco-jb) [inline contracts/pool_stable/src/contract.rs:503]: Good finding!

File: contracts/factory/src/contract.rs [hunk 4]
```diff
@@ -275,6 +277,78 @@ impl FactoryTrait for Factory {
...
+                &address,
+                &Symbol::new(&env, "query_pool_info_for_factory"),
+                Vec::new(&env),
+            );
+
+            let asset_addr = response.pool_response.asset_a.address.clone();
+            let bsset_addr = response.pool_response.asset_b.address.clone();
+
+            let (asset_amount_available_for_user, bsset_amount_available_for_user) =
+                get_token_amount_per_liquidity_pool(&env, &asset_addr, &bsset_addr, &sender);
```
Comment (ueco-jb) [inline contracts/factory/src/contract.rs:297]: This is wrong. You are providing current user's balance as amount of tokens in the liquidity pool?

File: contracts/factory/src/contract.rs [hunk 5]
```diff
@@ -275,6 +277,78 @@ impl FactoryTrait for Factory {
...
+    );
+
+    let bsset_amount = env.invoke_contract(
+        bsset_addr,
+        &Symbol::new(env, "balance"),
+        vec![env, sender.clone().into_val(env)],
+    );
+
+    (asset_amount, bsset_amount)
 }
```
Comment (ueco-jb) [inline contracts/factory/src/contract.rs:352]: To remove.

```suggestion
```

**Review by ueco-jb — COMMENTED** [review]
In order to create that portfolio correctly you need to add function which estimates amount of tokens to be withdrawn based on amount of LP share tokens.

```rust
 query_share { amount } 

let (pools, total_share) = pool_info();

let mut share_ratio = Decimal::zero();
if !total_share.is_zero() {
    share_ratio = Decimal::from_ratio(amount, total_share);
}
let amount: pool.amount * share_ratio;
```

Make this into a new query messge on both pool and pool stable. Add tests to confirm that.

File: contracts/factory/src/contract.rs [hunk 6]
```diff
@@ -275,6 +277,57 @@ impl FactoryTrait for Factory {
             .get(&DataKey::Config)
             .expect("Factory: No multihop present in storage")
     }
+
+    fn get_user_portfolio(env: Env, sender: Address) -> UserPortfolio {
```
Comment (ueco-jb) [inline contracts/factory/src/contract.rs:281]:
```suggestion
    fn query_user_portfolio(env: Env, sender: Address) -> UserPortfolio {
```

Plus add a boolean flag `staking: Bool`; query the staking information only if it is true.

Comment (gangov) [reply contracts/factory/src/contract.rs:281]: alright

File: contracts/factory/src/contract.rs (hunk 4 above)
Comment (gangov) [reply contracts/factory/src/contract.rs:297]: deleted

File: contracts/factory/src/contract.rs (hunk 6 above)
Comment (ueco-jb) [reply contracts/factory/src/contract.rs:281]: Just `staking: bool`

File: contracts/factory/src/tests/queries.rs [hunk 7]
```diff
@@ -390,3 +391,232 @@ fn test_queries_by_tuple_errors() {
...
+                        }
+                    )
+                }
+            ],
+            stake_portfolio: vec![
+                &env,
+                StakePortfolio {
+                    stake_token: stake_address,
+                    stakes: vec![&env] // no bonds, empty stake
+                }
```
Comment (ueco-jb) [inline contracts/factory/src/tests/queries.rs:516]: What's the point of this test if you didn't even stake? How do you know it's correct?

File: contracts/pool/src/tests/liquidity.rs [hunk 8]
```diff
@@ -698,3 +700,167 @@ fn withdraw_liqudity_below_min() {
     // Expecting min_a and/or min_b as huge bigger then available
     pool.withdraw_liquidity(&user1, &share_amount, &3000, &3000);
 }
+
+#[test]
+// todo add failing test cases as well
+// i.e. query_share with empty pool etc
```
Comment (ueco-jb) [inline contracts/pool/src/tests/liquidity.rs:706]:
```suggestion
```

File: contracts/pool/src/tests/liquidity.rs [hunk 9]
```diff
@@ -698,3 +700,167 @@ fn withdraw_liqudity_below_min() {
...
+fn query_share_valid_liquidity() {
```
Comment (ueco-jb) [inline contracts/pool/src/tests/liquidity.rs:707]: Add two more users who provide liquidity to this pool and verify them all.

File: contracts/pool/src/tests/liquidity.rs [hunk 10]
```diff
@@ -698,3 +700,167 @@ fn withdraw_liqudity_below_min() {
...
+                address: token2.address.clone(),
+                amount: 0i128
+            },
+            asset_lp_share: Asset {
+                address: share_token_address,
+                amount: 0i128
+            },
+            stake_address: pool_info_result.clone().stake_address,
+        }
+    );
```
Comment (ueco-jb) [inline contracts/pool/src/tests/liquidity.rs:793]: How is this even working?
You provided 150/200 tokens, withdraw 100/100 and now the pool is 0?

File: contracts/pool/src/tests/liquidity.rs [hunk 11]
```diff
@@ -698,3 +700,167 @@ fn withdraw_liqudity_below_min() {
...
+
+    let share_token_address = pool.query_share_token_address();
+    let token_share = token_contract::Client::new(&env, &share_token_address);
+
+    token1.mint(&user1, &1000);
+    assert_eq!(token1.balance(&user1), 1000);
+
+    token2.mint(&user1, &1000);
+    assert_eq!(token2.balance(&user1), 1000);
+
```
Comment (ueco-jb) [inline contracts/pool/src/tests/liquidity.rs:850]:
```suggestion
```

**Review by ueco-jb — CHANGES_REQUESTED** [review]

**Comment by ueco-jb** [comment]
Add at least one user to each of those testcases where you provide liquidity.

File: contracts/pool/src/tests/liquidity.rs (hunk 10 above)
Comment (gangov) [reply contracts/pool/src/tests/liquidity.rs]: we covered that in private

File: contracts/factory/src/tests/setup.rs [hunk 12]
```diff
@@ -7,13 +11,40 @@ pub mod lp_contract {
...
+) -> stake_contract::Client<'a> {
+    stake_contract::Client::new(
+        env,
+        &env.register_contract_wasm(Some(&stake_contract_address), stake_contract::WASM),
+    )
+}
+
+pub fn deploy_token_contract<'a>(env: &Env, admin: &Address) -> token_contract::Client<'a> {
+    token_contract::Client::new(env, &env.register_stellar_asset_contract(admin.clone()))
+}
```
Comment (ueco-jb) [inline contracts/factory/src/tests/setup.rs:46]: To remove. Those contracts are deployed by the factory...

File: contracts/factory/src/tests/queries.rs [hunk 13]
```diff
@@ -390,3 +394,278 @@ fn test_queries_by_tuple_errors() {
...
+    };
+
+    let lp_contract_addr = factory.create_liquidity_pool(
+        &admin.clone(),
+        &first_lp_init_info,
+        &String::from_str(&env, "Pool"),
+        &String::from_str(&env, "PHO/BTC"),
+    );
+
+    let first_lp_client = deploy_lp_contract(&env, lp_contract_addr.clone());
```
Comment (ueco-jb) [inline contracts/factory/src/tests/queries.rs:638]:
```suggestion
    let first_lp_client = lp_contract::Client::new(&env, &lp_contract_addr.clone());
```

File: contracts/pool/src/tests/liquidity.rs [hunk 14]
```diff
@@ -698,3 +700,281 @@ fn withdraw_liqudity_below_min() {
...
+                amount: 301
+            },
+            asset_b: Asset {
+                address: token2.address.clone(),
+                amount: 401
+            },
+            asset_lp_share: Asset {
+                address: share_token_address.clone(),
+                amount: 346i128
+            },
```
Comment (ueco-jb) [inline contracts/pool/src/tests/liquidity.rs:796]: That assertion makes sense if you compare it with the balance before withdrawing.

**Review by ueco-jb — COMMENTED** [review]
The same tests needs to be included in the pool stable implementation.

**Comment by gangov** [comment]
> The same tests needs to be included in the pool stable implementation.

yes, I was thinking that once we have the pool_stable tests enabled I can add them, as well as the other tests that are pending to be added. Does that sounds good?

File: contracts/pool/src/tests/liquidity.rs (hunk 14 above)
Comment (gangov) [reply contracts/pool/src/tests/liquidity.rs]: alright, made a query for the pool info before and after the withdrawal and now I'm comparing that. After the final withdrawal we assert that evyrhing in pool_info is at 0

File: contracts/factory/src/tests/queries.rs [hunk 15]
```diff
@@ -390,3 +393,298 @@ fn test_queries_by_tuple_errors() {
...
+        &admin.clone(),
+        &first_lp_init_info,
+        &String::from_str(&env, "Pool"),
+        &String::from_str(&env, "PHO/BTC"),
+    );
+
+    let first_lp_client = lp_contract::Client::new(
+        &env,
+        &env.register_contract_wasm(Some(&first_lp_contract_addr), lp_contract::WASM),
+    );
```
Comment (ueco-jb) [inline contracts/factory/src/tests/queries.rs:452]:
```suggestion
    let first_lp_client = lp_contract::Client::new(
        &env,
        &lp_contract_addr.clone(),
    );
```

File: contracts/factory/src/contract.rs [hunk 16]
```diff
@@ -275,6 +277,64 @@ impl FactoryTrait for Factory {
...
+                let stake_response: StakedResponse = env.invoke_contract(
+                    &response.pool_response.stake_address,
+                    &Symbol::new(&env, "query_staked"),
+                    vec![&env, sender.into_val(&env)],
+                );
+
+                // only stakes that the user has made
+                if !stake_response.stakes.is_empty() {
+                    stake_portfolio.push_back(StakePortfolio {
+                        stake_token: response.pool_response.stake_address,
```
[the remark this answers is not in the document]
Comment (gangov) [reply contracts/factory/src/contract.rs]: changed

**Comment by gangov** [comment]
rebased from main, will add the required tests now

File: contracts/pool/src/tests/liquidity.rs [hunk 17]
```diff
@@ -698,3 +702,253 @@ fn withdraw_liqudity_below_min() {
...
+            Asset {
+                address: token1.address.clone(),
+                amount: 0
+            },
+            Asset {
+                address: token2.address.clone(),
+                amount: 0
+            }
+        )
+    );
```
Comment (ueco-jb) [inline contracts/pool/src/tests/liquidity.rs:806]: Make multiple users bond at the same time, there's no point in bonding/unbonding each user separately.

**Comment by gangov** [comment]
pool_stable's `PoolResponse` struct has no `stake_address` field, wouldn't this be necessary?

**Comment by ueco-jb** [comment]
> pool_stable's PoolResponse struct has no stake_address field, wouldn't this be necessary?

yes

File: contracts/factory/src/tests/queries.rs [hunk 18]
```diff
@@ -390,3 +393,283 @@ fn test_queries_by_tuple_errors() {
...
+        }
+    );
+
+    second_stake_client.bond(&user_2, &223i128);
+
+    let second_portfolio = factory.query_user_portfolio(&user_2, &true);
+    assert_eq!(
+        second_portfolio,
+        UserPortfolio {
+            lp_portfolio: vec![&env,],
```
Comment (ueco-jb) [inline contracts/factory/src/tests/queries.rs:571]: Why in this case is `lp_portfolio` an empty vector?

File: contracts/pool_stable/src/tests/queries.rs [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+            Asset {
+                address: token1.address.clone(),
+                amount: 0
+            },
+            Asset {
+                address: token2.address.clone(),
+                amount: 0
+            }
+        )
+    );
```
Comment (ueco-jb) [inline contracts/pool_stable/src/tests/queries.rs:192]: You query using 0 as an argument, it doesn't make sense to include that.

**Review by ueco-jb — COMMENTED** [review]
Last thing I noticed we are missing a test where there are multiple liquidity pools. Please add one where there are at least two pools and both display shares within it correctly.

File: contracts/factory/src/tests/queries.rs (hunk 18 above)
Comment (gangov) [reply contracts/factory/src/tests/queries.rs]: we were querying for lp_share tokens only in the pool and not the ones that were sent by the user to the corresponding staking contract. Thanks

**Review by ueco-jb — COMMENTED** [review]
Looks good. Last comment - in tests where you test multiple pools I think it's fine if you try `with_stake` option just once, at the end to make sure. No need to test it every time, it only makes tests longer.
Other then that LGTM.

**Review by ueco-jb — APPROVED** [review]
LGTM
Good job

**Pull request merged**